docker/Dockerfile

```
@@ -91,117 +95,125 @@ RUN python /usr/local/generate_toolchain.py --prefix /usr
-RUN /usr/local/package/intel-compute-runtime.sh https://github.com/intel/compute-runtime/archive/21.31.20514.tar.gz \
-    https://patch-diff.githubusercontent.com/raw/intel/compute-runtime/pull/550.patch -- \
+RUN --mount=type=bind,source=usr/local/package/intel-compute-runtime.sh,target=/usr/local/package/intel-compute-runtime.sh \
+    /usr/local/package/intel-compute-runtime.sh https://github.com/intel/compute-runtime/archive/refs/tags/22.29.23750.tar.gz \
+    https://github.com/intel/compute-runtime/commit/78c96db2faacdc3406a24a91f017aeae40b7ca54.patch \
+    -- \
     -DSKIP_UNIT_TESTS=ON \
     -DIGC_DIR=/usr/local
 
-# Install OpenCL C and C++ headers
-COPY usr/local/package/opencl-headers.sh /usr/local/package/opencl-headers.sh
-RUN /usr/local/package/opencl-headers.sh https://github.com/KhronosGroup/OpenCL-Headers/archive/v2020.12.18.tar.gz
-RUN install-from-source https://github.com/KhronosGroup/OpenCL-CLHPP/archive/refs/tags/v2.0.16.tar.gz \
-    -DOPENCL_INCLUDE_DIR=/usr/local/include/CL \
-    -DOPENCL_LIB_DIR=/usr/local/lib \
-    -DBUILD_DOCS=OFF \
-    -DBUILD_EXAMPLES=OFF \
-    -DBUILD_TESTS=OFF
+# Install OpenCL C Headers
+RUN install-from-source https://github.com/KhronosGroup/OpenCL-Headers/archive/refs/tags/v2022.05.18.tar.gz
 
 # OpenCL loader library
 RUN install-package ruby
-RUN install-from-source https://github.com/OCL-dev/ocl-icd/archive/v2.2.14.tar.gz
-
-# OpenBLAS
-RUN install-package gcc-fortran
-COPY usr/local/package/openblas/${platform}.sh /usr/local/package/openblas.sh
-RUN /usr/local/package/openblas.sh https://github.com/xianyi/OpenBLAS/archive/v0.3.13.tar.gz
+RUN install-from-source https://github.com/OCL-dev/ocl-icd/archive/refs/tags/v2.3.1.tar.gz
 
 # Eigen3
-COPY usr/local/package/eigen3/deprecated-copy.patch /usr/local/package/deprecated-copy.patch
-COPY usr/local/package/eigen3/cxx17-overalign.patch /usr/local/package/cxx17-overalign.patch
-RUN install-from-source-with-patches https://gitlab.com/libeigen/eigen/-/archive/3.3.7/eigen-3.3.7.tar.bz2 \
-    /usr/local/package/deprecated-copy.patch \
-    /usr/local/package/cxx17-overalign.patch --
+RUN install-from-source https://gitlab.com/libeigen/eigen/-/archive/3.4.0/eigen-3.4.0.tar.gz
 
 # tcmalloc
-RUN install-from-source \
-    https://github.com/gperftools/gperftools/releases/download/gperftools-2.7/gperftools-2.7.tar.gz \
+RUN install-autotools-from-source https://github.com/gperftools/gperftools/releases/download/gperftools-2.10/gperftools-2.10.tar.gz \
     --with-tcmalloc-pagesize=64 \
     --enable-minimal
 
 # Protobuf
-RUN install-package protobuf
-RUN install-autotools-from-source https://github.com/protocolbuffers/protobuf/releases/download/v3.17.3/protobuf-cpp-3.17.3.tar.gz \
+RUN install-package protobuf && \
+    install-autotools-from-source https://github.com/protocolbuffers/protobuf/releases/download/v21.4/protobuf-cpp-3.21.4.tar.gz \
     --with-zlib=/usr/local \
     --with-protoc=/usr/sbin/protoc
```

With changes in `Neutron.cmake` to use DEPFILEs in `add_custom_command`s `ninja` will always think that the message protos are dirty and will always rebuild them.

The reason that `ninja` thinks they are dirty is because it treats the first specified output as the "primary" output and then gets upset that the primary output isnt the first output mentioned in the DEPFILE.

Unfortunately, `protobuf` uses and `std::unordered_map` to store output file names and so we are unable to rely on a consistent ordering of the outputs in the DEPFILE. This patch sorts the output files to give us that consistent order
```suggestion
RUN install-package protobuf && \
    install-from-source-with-patches cmake \
    https://github.com/protocolbuffers/protobuf/releases/download/v21.4/protobuf-cpp-3.21.4.tar.gz \
    https://gist.githubusercontent.com/TrentHouliston/f2e318ff0b96991a840778f10ac2dc63/raw/86bab8774fcb1f21b2bc8ed4fd35e9ecfddf3b5d/sort_manifest.patch \
    -- \
    -DWITH_PROTOC="/usr/sbin/protoc" \
    -Dprotobuf_BUILD_TESTS=OFF \
    -Dprotobuf_BUILD_LIBPROTOC=ON \
    -DBUILD_SHARED_LIBS=ON \
    -DZLIB_ROOT=${prefix}
```

---

Suggesting I use my own patch!
I'll make it a proper PR then link to that commit for the patch